packages/serverpod_client/lib/src/auth_key_provider.dart

```
@@ -9,28 +9,49 @@ abstract interface class ClientAuthKeyProvider {
-    // about to expire to avoid racing conditions when multiple endpoints try to
-    // refresh and one has already refreshed. Then skip refresh and return true.
-
-    // TODO: Use Completer to implement a lock under the future of the getter to
-    // avoid multiple concurrent refresh attempts.
-    return await notLockedAuthHeaderValue;
-  }
-
+abstract interface class RefresherClientAuthKeyProvider
+    implements ClientAuthKeyProvider {
   /// Refreshes the authentication key. If the refresh is successful, should
   /// return true to retry requests that failed due to authentication errors.
   /// Be sure to annotate the refresh endpoint with @unauthenticatedClientCall
   /// to avoid a deadlock on the [authHeaderValue] getter on refresh call.
   Future<bool> refreshAuthKey();
 }
+
+/// A [RefresherClientAuthKeyProvider] decorator that adds a mutex lock to
+/// prevent concurrent refresh calls. Actual auth header getter and refresh
+/// logic is delegated to the [_delegate] provider.
+class MutexRefresherClientAuthKeyProvider
+    implements RefresherClientAuthKeyProvider {
+  final RefresherClientAuthKeyProvider _delegate;
+
+  /// Creates a new [MutexRefresherClientAuthKeyProvider].
+  MutexRefresherClientAuthKeyProvider(this._delegate);
+
+  /// Shared future that serves as a lock to prevent concurrent refresh calls.
+  Future<bool>? _pendingRefresh;
+
+  @override
+  Future<String?> get authHeaderValue async {
+    await refreshAuthKey();
+    return _delegate.authHeaderValue;
+  }
+
+  /// Refreshes the authentication key with locking to prevent concurrent calls.
+  @override
+  Future<bool> refreshAuthKey() async {
+    final pendingRefresh = _pendingRefresh;
+    if (pendingRefresh != null) return pendingRefresh;
+
+    final refreshFuture = _delegate.refreshAuthKey();
+    _pendingRefresh = refreshFuture;
+
+    try {
+      return await refreshFuture;
+    } finally {
+      if (identical(_pendingRefresh, refreshFuture)) {

```

Suggestion: Throw state error on `!identical(...)`.

If this check fails that means our mutex it now working. We should probably throw a state error in that case since we might be messing up more things. Better to make us informed if that is the case.